internal/services/pam/pam_test.go

```diff
@@ -0,0 +1,596 @@
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"github.com/ubuntu/authd"
+	"github.com/ubuntu/authd/internal/brokers"
+	"github.com/ubuntu/authd/internal/services/pam"
+	"github.com/ubuntu/authd/internal/testutils"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+var (
+	generatedID string
+	testClient  authd.PAMClient
+)
+
+// Used for TestGetAuthenticationModes and TestSelectAuthenticationMode.
+var requiredEntries = "required:entry_type,other_entry_type"
+var optionalEntries = "optional:entry_type,other_entry_type"
+var optional = "optional"
+var requiredEntry = &authd.UILayout{
+	Type:    "required-entry",
+	Label:   &optional,
+	Button:  &optional,
+	Wait:    &optional,
+	Entry:   &requiredEntries,
+	Content: &optional,
+}
+var optionalEntry = &authd.UILayout{
+	Type:  "optional-entry",
+	Entry: &optionalEntries,
+}
+var emptyType = &authd.UILayout{
+	Type:  "",
+	Entry: &requiredEntries,
+}
+
+func TestAvailableBrokers(t *testing.T) {
+	t.Parallel()
+
+	abResp, err := testClient.AvailableBrokers(context.Background(), &authd.Empty{})
+	require.NoError(t, err, "AvailableBrokers should not return an error, but did")
+
+	got := abResp.GetBrokersInfos()
+	for _, broker := range got {
+		broker.Id = broker.Name + "_ID"
+	}
+	want := testutils.LoadWithUpdateFromGoldenYAML(t, got)
+	require.Equal(t, want, got, "AvailableBrokers returned unexpected brokers")
+}
+
+func TestSetDefaultBrokerForUser(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		username string
+		noBroker bool
+
+		wantErr bool
+	}{
+		"Set default broker for existing user": {username: "user"},
+
+		"Error when username is empty":     {wantErr: true},
+		"Error when broker does not exist": {username: "user", noBroker: true, wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			abResp, err := testClient.AvailableBrokers(context.Background(), &authd.Empty{})
+			require.NoError(t, err, "Setup: failed to get available brokers for tests")
+
+			wantID := abResp.GetBrokersInfos()[0].GetId()
+			if tc.noBroker {
+				wantID = "does not exist"
+			}
+
+			sdbfuReq := &authd.SDBFURequest{
+				BrokerId: wantID,
+				Username: tc.username,
+			}
+			_, err = testClient.SetDefaultBrokerForUser(context.Background(), sdbfuReq)
+			if tc.wantErr {
+				require.Error(t, err, "SetDefaultBrokerForUser should return an error, but did not")
+				return
+			}
+			require.NoError(t, err, "SetDefaultBrokerForUser should not return an error, but did")
+
+			gotResp, _ := testClient.GetPreviousBroker(context.Background(), &authd.GPBRequest{Username: tc.username})
+			require.Equal(t, wantID, gotResp.GetPreviousBroker(), "SetDefaultBrokerForUser did not set the correct broker for the user")
+		})
+	}
+}
+
+func TestGetPreviousBroker(t *testing.T) {
+	t.Parallel()
+
+	_, err := testClient.SetDefaultBrokerForUser(context.Background(), &authd.SDBFURequest{
```

Comment:
I would even do the no broker/broker with the same user:
GetPreviousBroker -> nothing
SetDefaultBroker…
GetPreviousBroker -> oh, something!

If we have 2 different users, I feel those are two independent subtests so that only one fails. This could be even better, but unsure if you want to add that setup. This would be even better IMHO. However, I agree Set/Get are a little bit similar, so it’s detabatable to get to this subtest.